firewood/src/merkle/stream.rs

```diff
@@ -7,399 +7,471 @@ use crate::{
-                    .map_err(|e| api::Error::InternalError(Box::new(e)))
-                    .transpose();
+                let partial_key = match child.inner() {
+                    NodeType::Branch(branch) => &branch.path,
+                    NodeType::Leaf(leaf) => &leaf.path,
+                    NodeType::Extension(_) => {
+                        panic!("extension nodes shouldn't exist")
+                    }
+                };
 
-                Poll::Ready(next)
+                let (comparison, new_unmatched_key_nibbles) =
+                    is_prefix(partial_key.iter(), unmatched_key_nibbles);
+                unmatched_key_nibbles = new_unmatched_key_nibbles;
+
+                match comparison {
+                    Ordering::Less => {
+                        // `child` is before `key`.
+                        return Ok(MerkleNodeStreamState::Initialized { iter_stack });
+                    }
+                    Ordering::Equal => {
+                        // `child` is a prefix of `key`.
+                        matched_key_nibbles.extend(partial_key.iter().copied());
+                        node = child;
+                        node_addr = child_addr;
+                    }
+                    Ordering::Greater => {
+                        // `child` is after `key`.
+                        let child = merkle
+                            .get_node(child_addr)
+                            .map_err(|e| api::Error::InternalError(Box::new(e)))?;
+                        iter_stack.push(IterationNode::Unvisited {
+                            key: matched_key_nibbles
+                                .iter()
+                                .copied()
+                                .chain(partial_key.iter().copied())
+                                .collect(),
+                            node: child,
+                        });
+
+                        return Ok(MerkleNodeStreamState::Initialized { iter_stack });
+                    }
+                }
             }
-        }
+            NodeType::Leaf(leaf) => {
+                let (comparison, _) = is_prefix(leaf.path.iter(), unmatched_key_nibbles);
+
+                match comparison {
+                    Ordering::Less | Ordering::Equal => {}
+                    Ordering::Greater => {
+                        let node = merkle
+                            .get_node(node_addr)
+                            .map_err(|e| api::Error::InternalError(Box::new(e)))?;
+                        // The leaf's key > `key`, so we can stop here.
+                        iter_stack.push(IterationNode::Unvisited {
+                            key: matched_key_nibbles
+                                .iter()
+                                .copied()
+                                .chain(leaf.path.iter().copied())
+                                .collect(),
+                            node,
+                        });
+                    }
+                }
```

Comment:
I'm surprised clippy didn't catch this, but it's just an if-statement

```suggestion
                if is_prefix(leaf.path.iter(), unmatched_key_nibbles).0 == Ordering::Greater {
                   let node = merkle
                        .get_node(node_addr)
                        .map_err(|e| api::Error::InternalError(Box::new(e)))?;

                    // The leaf's key > `key`, so we can stop here.
                    let key = matched_key_nibbles
                        .iter()
                        .copied()
                        .chain(leaf.path.iter().copied())
                        .collect();
                    
                    iter_stack.push(IterationNode::Unvisited {
                        key,
                        node,
                    });
                }
```